### PR TITLE
ALL-3936/JavaDoc

### PR DESCRIPTION
**Pull request by Mecharyry**
## Problem
As per issue #40 JavaDoc public APIs. We have very little information on any of our public APIs to aid clients of this library.

## Solution
Add JavaDoc to `Player` and `PlayerBuilder` the two public APIs.

### Test(s) added 
No, adding JavaDoc only.

### Screenshots
No UI changes.

### Paired with 
@ouchadam

### CODE REVIEW

File: core/src/main/java/com/novoda/noplayer/Player.java [hunk 1]
```diff
@@ -14,38 +14,121 @@
 
 public interface Player extends PlayerState {
 
+    /**
```
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/Player.java]: I think that 
```
/**
* Plays content of a prepared {@link Player}.
* @see Player.PreparedListener
*/
``` 
Would read much better. also for the others

File: core/src/main/java/com/novoda/noplayer/Player.java [hunk 2]
```diff
@@ -14,38 +14,121 @@
...
     void stop();
 
+    /**
+     * Stops playback of content and drops all internal resources. The instance of {@link Player} should not be
+     * used after calling {@link Player#release()}.
+     */
     void release();
 
+    /**
+     * Loads the video content and triggers the {@link Player.PreparedListener}.
```
Comment (juankysoriano) [inline core/src/main/java/com/novoda/noplayer/Player.java]:
```
/**
* Loads video content and prepares the player. 
* @see Player.PreparedListener 
*/
```

Same for the other

File: core/src/main/java/com/novoda/noplayer/player/PlayerBuilder.java [hunk 3]
```diff
@@ -52,12 +91,26 @@ public PlayerBuilder withPriority(PlayerType playerType, PlayerType... playerTyp
...
         return this;
     }
 
-    public Player build(Context context) {
+    /**
+     * Builds a new Player instance.
+     *
+     * @param context
+     * @return a Player instance.
+     * @throws UnableToCreatePlayerException thrown when the configuration is not supported and there is no way to recover.
```
Comment (zegnus) [inline core/src/main/java/com/novoda/noplayer/player/PlayerBuilder.java]: :)

**Comment by Mecharyry** [comment]
I guess quite a few of the `Player` methods can throw, if the client doesn't make sure the `Player` is prepared before calling some methods.

**Comment by zegnus** [comment]
@Mecharyry do you think the client might benefit on knowing that not only the method won't work properly, but that an exception will be thrown

**Comment by Mecharyry** [comment]
I'm not sure. We highlight that the `Player` has to be prepared before calling the method. I think adding the `exception` would just be repeating ourselves. I'm not sure though.

**Comment by zegnus** [comment]
Do we have any disadvantage documenting the exception?

**Comment by Mecharyry** [comment]
No, we should document it I think. Better to be overly verbose on a JavaDoc than miss something out and have issues on the library.

**Comment by zegnus** [comment]
👍

**Comment by Mecharyry** [comment]
@zegnus I've tried adding the throws for all of the methods on the `Player` but I'm not sure this is doable. Here's my reasoning.

- We are throwing an `UncheckedException` which we aren't expecting clients to catch and handle. It represents a failure of state based on how they are using the library. 
- The underlying libraries although based on the same interface could be initialised in very different ways and what may cause a crash in one does not necessarily mean it will crash in the other.

**Comment by zegnus** [comment]
@Ryan I'll reply in order:

- The doc is not for expecting a client to handle it, but to be aware what happens with that method. Although the java conventions doesn't mention that you have to document all the unchecked exceptions, they are properly documented in the majority of the Java Standard Libraries (example: https://docs.oracle.com/javase/7/docs/api/java/util/List.html#get(int))

- As a user, if I code against an interface I would expect to have to call the methods in the same order regardless of the internal implementation. Even if its not the case, makes the public API much simpler to just provide one way to perform an action. You can also elaborate in which case an exception is thrown in the doc

File: core/src/main/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerImpl.java [hunk 4]
```diff
@@ -261,33 +260,28 @@ public void detach(PlayerView playerView) {
...
     @Override
-    public boolean selectAudioTrack(PlayerAudioTrack audioTrack) {
+    public boolean selectAudioTrack(PlayerAudioTrack audioTrack) throws IllegalStateException {
         return mediaPlayer.selectAudioTrack(audioTrack);
     }
 
     @Override
     public boolean showSubtitleTrack(PlayerSubtitleTrack subtitleTrack) {
-        SubtitlesError subtitlesError = new SubtitlesError("Subtitles not implemented for Android Media Player", new IllegalStateException());
-        listenersHolder.getErrorListeners().onError(this, subtitlesError);
```
Comment (zegnus) [inline core/src/main/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerImpl.java]: Are you sure we want to remove the error report in these methods?

Comment (Mecharyry) [inline core/src/main/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerImpl.java]: Yes. After talking to Adam, it's not worth reporting in this way. It's not supported in `MediaPlayer`but it should be in the future. I would like to have a wider array of `Capabilities` that `Player`s can support and then we select a `Player` based on that.

**Comment by zegnus** [comment]
We'll move some classes to an `internal` package and update documentation on those so its clear from the client

File: core/src/test/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerFacadeTest.java [hunk 5]
```diff
@@ -335,6 +336,18 @@ public void givenMediaPlayerIsNotPlaying_whenPausing_thenDoesNotPausesMediaPlaye
     }
 
     @Test
+    public void givenMediaPlayerIsNotInPlaybackState_whenPausing_thenThrowsIllegalStateException() {
```
Comment (juankysoriano) [inline core/src/test/java/com/novoda/noplayer/mediaplayer/AndroidMediaPlayerFacadeTest.java]: 👍

**Pull request merged**